castervoice/lib/migration.py

```diff
@@ -0,0 +1,79 @@
+import shutil
+
+import six
+
+from castervoice.lib import settings
+from castervoice.lib.ctrl.mgr.loading.load.content_root import ContentRoot
+from castervoice.lib.merge.selfmod.sm_config import SelfModStateSavingConfig
+
+if six.PY2:
+    from castervoice.lib.util.pathlib import Path
+else:
+    from pathlib import Path  # pylint: disable=import-error
+
+
+class UserDirUpdater(object):
+
+    def __init__(self, user_dir):
+        self.user_dir = user_dir
+        self.root_user_content_package_name = ContentRoot.USER_DIR
+
+    def create_user_dir_directories(self):
+        for directory in ["data", "sikuli", "settings"]:
+            Path(self.user_dir).joinpath(directory).mkdir(parents=True, exist_ok=True)
+
+    def update_user_dir_packages_to_v1_7_0(self):
+        """
+        Migrates Caster user dir from v1.0.0 structure to v1.7.0+ structure.
+        If used for first time, just sets up v1.7.0+ structure.
+        """
+        self._make_root_user_content_package()
+        for directory in ["rules", "transformers", "hooks"]:
+            self._move_legacy_package(directory)
+            self._make_user_content_package(directory)
+
+    def update_bringme_toml_to_v1_7_0(self):
+        user_dir_path = settings.settings(["paths", "USER_DIR"])
+        bringme_config_path = settings.settings(["paths", "SM_BRINGME_PATH"])
+        bringme_config = SelfModStateSavingConfig(bringme_config_path)
+        bringme_config.load()
+        directory_dict = bringme_config.get("folder")
+
+        if ContentRoot.USER_DIR not in directory_dict["caster rules"]:
```

Comment:
`if directory_dict is not None and ContentRoot.USER_DIR not in directory_dict["caster rules"]:`

checking directory_dict if is None which would be the case if the user directory hasn't been generated yet on 1st run.

Reply:
https://github.com/dictation-toolbox/Caster/commit/0470ec9ed58a3a64f9652c331868c30b108b9be5